index.js

```diff
@@ -30,22 +30,40 @@ function compose (middleware) {
    * @api public
    */
 
-  return function (context, next) {
-    // last called middleware #
-    let index = -1
-    return dispatch(0)
-    function dispatch (i) {
-      if (i <= index) return Promise.reject(new Error('next() called multiple times'))
-      index = i
-      const fn = middleware[i] || next
-      if (!fn) return Promise.resolve()
-      try {
-        return Promise.resolve(fn(context, function next () {
-          return dispatch(i + 1)
-        }))
-      } catch (err) {
-        return Promise.reject(err)
+  middleware[Symbol.iterator] = function () {
+    let self = this
+    let i = 0
+    let c
+    let n
```

Comment:
Could you make these variable names more descriptive, or add some comments? Right now it's a bit hard to figure out what's going on here.